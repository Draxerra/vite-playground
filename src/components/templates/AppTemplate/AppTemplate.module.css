.app-template {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  min-block-size: 100%;
  min-inline-size: 100%;
  padding-block: var(--block-spacing);
  padding-inline: var(--inline-spacing);
}

.app-template__main {
  display: grid;
  place-content: center;

  flex: 1;
}
