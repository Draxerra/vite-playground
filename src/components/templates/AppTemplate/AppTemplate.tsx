import React, { FC, lazy, Suspense } from "react";
import { BrowserRouter as Router, Routes, Route } from "react-router-dom";

import styles from "./AppTemplate.module.css";

import AppNav from "~/components/organisms/AppNav";
import AppHome from "~/components/pages/AppHome";

const AppContact = lazy(() => import("~/components/pages/AppContact"));
const renderLoader = () => <div>Loading...</div>;

const AppTemplate: FC = () => (
  <Router>
    <div className={styles.appTemplate}>
      <header>
        <h1 className="sr-only">Vite React Test App</h1>
        <AppNav />
      </header>
      <main className={styles.appTemplatePage}>
        <Routes>
          <Route path="/" element={<AppHome />} />
          <Route
            path="contact"
            element={
              <Suspense fallback={renderLoader()}>
                <AppContact />
              </Suspense>
            }
          />
        </Routes>
      </main>
      <footer></footer>
    </div>
  </Router>
);

export default AppTemplate;
