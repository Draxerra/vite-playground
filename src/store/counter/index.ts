import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

export const asyncIncrement = createAsyncThunk(
  "counter/asyncIncrement",
  async () => {
    await new Promise((resolve) => setTimeout(resolve, 5000));
    return 1;
  }
);

export const counter = createSlice({
  name: "counter",
  initialState: {
    value: 0,
  },
  reducers: {
    increment: (state) => {
      state.value += 1;
    },
    decrement: (state) => {
      state.value -= 1;
    },
    incrementByAmount: (state, action: PayloadAction<number>) => {
      state.value += action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(asyncIncrement.fulfilled, (state, action) => {
      state.value += action.payload;
    });
  },
});

export const { increment, decrement, incrementByAmount } = counter.actions;

export default counter.reducer;
