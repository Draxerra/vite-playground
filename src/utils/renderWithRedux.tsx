import React, { FC } from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";

type RenderParams = Parameters<typeof render>;
type Render = ReturnType<typeof render>;
type StoreParams = Parameters<typeof configureStore>;
type Store = ReturnType<typeof configureStore>;

type RenderWithReduxOptions = {
  preloadedState?: StoreParams[0]["preloadedState"];
  reducers: StoreParams[0]["reducer"];
  store?: Store;
} & RenderParams[1];

const renderWithRedux: (
  ui: RenderParams[0],
  opts: RenderWithReduxOptions
) => Render = (
  ui,
  {
    preloadedState = {},
    reducers,
    store = configureStore({ reducer: reducers, preloadedState }),
    wrapper: WrapperElement,
    ...renderOptions
  }
) => {
  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>
      {WrapperElement ? <WrapperElement>{children}</WrapperElement> : children}
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

export * from "@testing-library/react";
export { renderWithRedux };
