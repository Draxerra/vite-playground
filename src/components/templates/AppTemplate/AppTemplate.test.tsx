import React from "react";
import AppTemplate from "./AppTemplate";
import { renderWithRedux, screen, waitFor } from "~/utils/renderWithRedux";

import counter from "~/store/counter";

describe("App Template", () => {
  afterEach(() => {
    window.history.pushState({}, "Home Page", "/");
  });

  it("should match snapshot", () => {
    const template = renderWithRedux(<AppTemplate />, {
      reducers: { counter },
    }).asFragment();
    expect(template).toMatchSnapshot();
  });

  it("should display loading page when url is /contact", () => {
    window.history.pushState({}, "Contact Page", "/contact");
    renderWithRedux(<AppTemplate />, {
      reducers: { counter },
    });
    const loading = screen.getByText("Loading...");
    expect(loading).toBeInTheDocument();
  });

  it("should display contact page when url is /contact", async () => {
    window.history.pushState({}, "Contact Page", "/contact");
    renderWithRedux(<AppTemplate />, {
      reducers: { counter },
    });
    await waitFor(() => {
      const contact = screen.getByText("Contact Page");
      expect(contact).toBeInTheDocument();
    });
  });
});
