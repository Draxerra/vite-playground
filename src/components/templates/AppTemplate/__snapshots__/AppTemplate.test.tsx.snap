// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Template should match snapshot 1`] = `
<DocumentFragment>
  <main
    class="appTemplate"
  >
    <nav
      class="appNav"
    >
      <a
        aria-current="page"
        aria-label="Navigate to Home Page"
        class="appLinkButton appLinkButtonActive appButton"
        href="/"
      >
        Home
      </a>
      <a
        aria-label="Navigate to Contact Page"
        class="appLinkButton appButton"
        href="/contact"
      >
        Contact
      </a>
    </nav>
    <section
      class="appTemplatePage"
    >
      <div
        class="appCounter"
      >
        <span
          aria-live="polite"
          class="appCounterText fs-600"
          role="region"
        >
          Current Count: 0
        </span>
        <button
          aria-label="Increment counter"
          class="appButton"
        >
          Increment
        </button>
        <button
          aria-label="Decrement counter"
          class="appButton"
        >
          Decrement
        </button>
      </div>
    </section>
  </main>
</DocumentFragment>
`;
